### PR TITLE
jl2xxx: add driver for jl2101 ethernet phy

### PR DESCRIPTION
**Pull request by 7Ji**
This ports the JL2xxx PHY driver used in LubanCatWrt's Rockchip 5.4 kernel for their DoorNet2 router development board to our 4.9 linux-amlogic kernel.

The driver is not directly available at JLsemi's [support page](https://www.jlsemi.com.cn/support.html) as they expect you to contact them when clicking their "Download" button. And I doubt I can get the driver successfully as an ordinary user, so decided to do this in a more dirty way.

This driver is available as patches at LubanCat's LubanCatWrt (their OpenWrt fork) repo, in the [rockchip 5.4 patches directory](https://github.com/LubanCat/LubanCatWrt/tree/master/target/linux/rockchip/patches-5.4), as it is an OpenWrt fork and this is placed in the build environment it inherits OpenWrt build system's [GPLv2 license](https://openwrt.org/license) and we should be able to use it in CoreELEC. 

There's other known projects that also ported this driver, such as [coolsnowwolf'lede](https://github.com/coolsnowwolf/lede/pull/9563) (another OpenWrt fork), and [unifreq's linux-5.15.y](https://github.com/unifreq/linux-5.15.y/commit/17deb8c7f3175fe3f8feb870fbbe79297e0501e1) (a fork of chewitt's linux that's used for some Armbian and OpenWrt forks). And since the source driver itself is used for a router product I believe it is more than reliable for our Kodi usage.

In summary, this PR ports the following two patches from LubanCatWrt's Rockchip 5.4 kernel:
 - https://github.com/LubanCat/LubanCatWrt/blob/master/target/linux/rockchip/patches-5.4/602-fix-kernel-stmmac-drivers-for-jl2101.patch
 - https://github.com/LubanCat/LubanCatWrt/blob/master/target/linux/rockchip/patches-5.4/602-fix-kernel-stmmac-drivers-for-jl2101.patch

And partially the following commit from upstream Linux (it's added to jl2xxx's header instead of the actual phy.h to avoid breaking other things):
 - https://github.com/torvalds/linux/commit/aa2af2eb447c9a21c8c9e8d2336672bb620cf900

This is tested on my HK1Box (S905X3) with a JL2101 PHY with sm1_s905x3_4g_1gbit.dtb with the following custom pressure test for 12 continous hours as there's no other more reliable tool (iperf3) for the pressure test
````
mkdir -p /tmp/pressure
cd /tmp/pressure
while true; do  wget http://192.168.7.1:8001/1Gtestfile; rm *; done
````
There's no kernel error message during the test. Its speed is capped around +900Mbps during the test, and is still around 900Mbps at the end of the test, and the PHY chip is not even warm, so I think it can handle the typical Kodi usage on CoreELEC very easily.

To enable this driver, the config ``CONFIG_JLSEMI_JL2XX1_PHY`` should be set to module(``=m``), this can be added directly in ``linux.aarch64.conf`` with a line ``CONFIG_JLSEMI_JL2XX1_PHY=m`` as all other deps are already built as modules

### CODE REVIEW

File: drivers/net/ethernet/stmicro/stmmac/stmmac_main.c [hunk 1]
```diff
@@ -1658,6 +1658,8 @@ static int stmmac_init_dma_engine(struct stmmac_priv *priv)
 	if (priv->extend_desc && (priv->mode == STMMAC_RING_MODE))
 		atds = 1;
 
+	msleep(1500);
```
Comment (Portisch) [inline drivers/net/ethernet/stmicro/stmmac/stmmac_main.c:1661]: Please cover such changes in main driver by config flag `JLSEMI_JL2XX1_PHY`
Such delay aren't good at all. Maybe there is another way to not affect other devices?

Comment (7Ji) [reply drivers/net/ethernet/stmicro/stmmac/stmmac_main.c:1661]: Let me try to remove the stmmac sections and do some tests to verify if it is really needed

**Comment by Portisch** [comment]
The `iperf` tools are available by install the `network-tools` addon from CE repo.

**Comment by 7Ji** [comment]
So the driver can't be loaded correctly if the system boots too quickly (here it means the normal CE boot speed), and will either generate lots of dmesg spams and softlocks the kernel or render the ethernet useless and only gets eth0 a 169 ip. On the first boot after burning CoreELEC though it can be loaded successfully, since it is slow to populate the storage partition. I think I can add the delay to the jl2xxx driver itself instead of to the stmmac driver.

**Comment by Portisch** [comment]
Maybe better, yes if it works. Is it not possible to make a single module driver and load it by event like a udev rule?

**Comment by 7Ji** [comment]
Yeah that would be a better idea, since I can't find a method to implement it in kernel while keeps the impact to other devices as minimum as possible: the driver expects certain changes in stmmc while it is nearly impossible to judge whether those operations are needed at the stage of stmmc initialization (kernel doesn't even know if there's phy). I'll close this and try to migrate the driver to a package instead.

**Comment by Portisch** [comment]
Maybe somehow it can be event loaded by the phy ID:
https://discourse.coreelec.org/t/report-about-amlogic-s905x4-hk1-rbox-x4/18603/172?u=portisch

**Pull request closed without merging**